airflow/settings.py

```diff
@@ -51,6 +52,36 @@
 log.info("Configured default timezone %s" % TIMEZONE)
 
 
+class RetryingQuery(Query):
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        try:
+            self.max_tries = conf.getint('core', 'SQL_ALCHEMY_STATEMENT_MAX_RETRIES')
+        except conf.AirflowConfigException:
+            self.max_tries = 10
+
+        try:
+            self.max_retry_time_seconds = conf.getint('core', 'SQL_ALCHEMY_STATEMENT_MAX_RETRY_SECONDS')
+        except conf.AirflowConfigException:
+            self.max_retry_time_seconds = 30
+
+    def __iter__(self):
+        try_number = 1
+        while try_number <= self.max_tries:
+            try:
+                return super().__iter__()
+
+            except Exception as ex:
+                log.warning(f'Try {try_number}/{self.max_tries} failed to perform db action.', exc_info=ex)
+                time.sleep(min(float(self.max_retry_time_seconds), 0.1 * (1 << try_number)))
+
+            try_number += 1
+
+        raise Exception(f'Failed to perform db action after {self.max_tries} attempts.')
```

Comment:
we should keep around the last exception so that we aren't just throwing away the exception info here. Ideally RetryingQuery with num_retries = 0 shouldn't appear any different from the normal Query

Reply:
if there is an exception in the last run, it will be logged in the `except` block above though?

Reply:
Right but it won't be bubbled up. We shouldn't just be catching exceptions and rethrowing them. The caller might handle the specific exception or care about the exception info